rest/client/iter.go

```diff
@@ -0,0 +1,54 @@
+package client
+
+// Iter defines an iterator type that should be used when implementing list methods. It's intended to
+// be embedded in a domain specific iterator struct.
+type Iter struct {
+	query Query
+
+	page    ListResponse
+	current interface{}
+	results []interface{}
+
+	err error
+}
+
+// Query defines a closure that domain specific iterators must implement. The implementation should
+// include a call to the API and should return the API response with a separate slice of the results.
+type Query func(string) (ListResponse, []interface{}, error)
+
+// GetIter returns a new initialized iterator. This method automatically makes the first query to
+// populate the results. List methods should use this helper method when building domain specific
+// iterators.
+func GetIter(url string, query Query) *Iter {
+	it := &Iter{query: query}
+	it.page, it.results, it.err = it.query(url)
+	return it
+}
+
+// Next moves the iterator to the next result.
+func (it *Iter) Next() bool {
+	if len(it.results) == 0 && it.page.NextPageURL() != "" {
+		it.page, it.results, it.err = it.query(it.page.NextPageURL())
+	}
+	if it.err != nil || len(it.results) == 0 {
+		return false
+	}
+	it.current = it.results[0]
+	it.results = it.results[1:]
+	return true
+}
+
+// Page returns the current results page in the form of an API response.
+func (it *Iter) Page() ListResponse {
+	return it.page
+}
+
+// Current returns the result that the iterator is currently pointing to.
+func (it *Iter) Current() interface{} {
```

Comment:
I don't love this returning an empty interface vs it returning the explicit type. I get there is a desire to have a "common" iter but I rather not have to type assert everywhere

Reply:
see [here](https://github.com/polygon-io/client-go/pull/13/files#diff-c2e88127eb60436aa954e3dd31a34149c0103930e39213858a45c48ce3c04a0eR17). this method is only meant for use by domain specific iters and users should call `iter.Quote()` instead

Reply:
Is there danger of that assertion failing? If not then this is fine

Reply:
if someone were to call `iter.Current()` and do the type assertion themselves (instead of `iter.Quote()` which handles nil) before `iter.Next()`, it'll return nil which would make the assertion fail. seems like it would be a deliberate choice by a user though. it's done that way so that people don't have to handle the first item in the list before starting the loop